lbry/file_analysis.py

```diff
@@ -0,0 +1,338 @@
+import asyncio
+import json
+import logging
+import os
+import pathlib
+import re
+import shlex
+import shutil
+
+from lbry.conf import TranscodeConfig
+
+log = logging.getLogger(__name__)
+
+
+class VideoFileAnalyzer:
+    @staticmethod
+    def _matches(needles: list, haystack: list):
+        for needle in needles:
+            if needle in haystack:
+                return True
+        return False
+
+    async def _execute(self, command, arguments):
+        args = shlex.split(arguments)
+        process = await asyncio.create_subprocess_exec(self._conf.ffmpeg_folder + command, *args,
+                                                       stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
+        stdout, stderr = await process.communicate()  # returns when the streams are closed
+        return stdout.decode() + stderr.decode(), process.returncode
+
+    async def _verify_executable(self, name):
+        try:
+            version, code = await self._execute(name, "-version")
+        except Exception as e:
+            log.warning("Unable to run %s, but it was requested. Message: %s", name, str(e))
+            code = -1
+            version = ""
+        if code != 0 or not version.startswith(name):
+            raise FileNotFoundError(f"Unable to locate or run {name}. Please install FFmpeg "
+                                    f"and ensure that it is callable via PATH or conf.ffmpeg_folder")
+        return version
+
+    async def _verify_ffmpeg_installed(self):
+        if self._ffmpeg_installed:
+            return
+        await self._verify_executable("ffprobe")
+        version = await self._verify_executable("ffmpeg")
+        log.debug("Using %s at %s", version.splitlines()[0].split(" Copyright")[0],
+                  shutil.which(f"{self._conf.ffmpeg_folder}ffmpeg"))
+        self._ffmpeg_installed = True
+
+    def __init__(self, conf: TranscodeConfig):
+        self._conf = conf
+        self._available_encoders = ""
+        self._ffmpeg_installed = False
```

Comment:
`__init__` should always go at the top of the class unless there is some very specific reason it shouldn't / can't.